dascore/proc/correlate.py

```diff
@@ -3,143 +3,204 @@
+    [`Patch.correlation_shift`](`dascore.proc.correlate.correlate_shift`)
+    is useful to undo dft artefacts after the idft is applied.
+
+    While a 2D patch is required for input, a 3D patch is returned where the
+    3rd dimesion cooresponds to the source rows/columns. For the case of a
+    single source, the [`Patch.squeeze`](`dascore.Patch.squeeze`) method
+    can be helpful to remove length 1 dimensions.
 
     Parameters
     ----------
     patch : PatchType
         The input data patch to be cross-correlated. Must be 2-dimensional.
-    lag :
-        An optional argument to save only certain lag times instead of full
-        output.
+        The patch can be in time or frequency domains.
     samples : bool, optional (default = False)
         If True, the argument specified in kwargs refers to the *sample* not
         value along that axis. See examples for details.
     **kwargs
-        Additional arguments to specify cross correlation dimension and the
-        master source, to which
-        we cross-correlate all other channels/time samples.
+        Specifies correlation dimension and the master source(s), to which
+        we want to cross-correlate all other channels/time samples.If the
+        master source is an array, the function will compute cc for all
+        the posible pairs.
 
     Examples
     --------
     >>> import dascore as dc
     >>> from dascore.units import m, s
 
-    >>> patch = dc.get_example_patch()
-
+    >>> # Get a patch composed of sin waves whose correlation results
+    >>> # can easily be checked.
+    >>> patch = dc.get_example_patch(
+    ...     "sin_wav",
+    ...     sample_rate=100,
+    ...     frequency=range(10, 20),
+    ...     duration=5,
+    ...     channel_count=10,
+    ... ).taper(time=0.5).set_units(distance='m')
+    >>>
     >>> # Example 1
     >>> # Calculate cc for all channels as receivers and
-    >>> # the 10 m channel as the master channel.
-    >>> cc_patch = patch.correlate(distance = 10 * m)
-
+    >>> # the 10 m channel as the master channel. Squeeze the output
+    >>> # so the returned patch is 2D.
+    >>> cc_patch = patch.correlate(distance = 10 * m).squeeze()
+    >>>
     >>> # Example 2
-    >>> # Calculate cc within (-2,2) sec of lag for all channels as receivers and
-    >>> # the 10 m channel as the master channel. The new patch has dimensions
-    >>> # (lag_time, distance)
-    >>> cc_patch = patch.correlate(distance = 10 * m, lag = 2 * s)
-
+    >>> # Get cc within (-2,2) sec of lag for all channels as receivers
+    >>> # and the 10 m channel as the master channel. The new patch has dimensions
+    >>> # (lag_time, distance, source_distance)
+    >>> cc_patch = (
+    ...     patch.correlate(distance = 10 * m)
+    ...     .select(lag_time=(-2, 2))
+    ... )
+    >>>
     >>> # Example 3
     >>> # Use 2nd channel (python is 0 indexed) along distance as master channel
     >>> cc_patch = patch.correlate(distance=1, samples=True)
```

Comment:
It'd be great to adjust this example to showcase how one can skip receivers.

```python
    >>> # Example 3
    >>> #  Correlate along distance having the 5th channel (python is 0 indexed) as a master channel
    >>> #  and skip every other channel for memory efficiency 
    >>> cc_patch = patch.decimate(distance=2).correlate(distance=4, samples=True)
```

Reply:
Sorry for back and forth on this but wouldn't be actually better to use select instead of decimate?

```python
distance = patch.get_array("distance")
cc_patch = patch.select(distace=distance[::2]).correlate(distance=2, samples=True)
```

Also, I think there is an issue with `Patch.select` as it does not actually select when an array is passed. I'm on the master branch and refreshed my installation.
```python
import dascore as dc

pa = dc.get_example_patch()

distance = pa.get_array("distance")
selected_pa = pa.select(distace=distance[::2])

assert pa.shape == (300,2000)
assert len(distance) == 300
assert len(distance[::2]) == 150
assert selected_pa.shape == (150,2000) # AssertionError here
```

Reply:
> wouldn't be actually more efficient to use select here instead of decimate?

I think both methods would work in different cases where aliasing may or may not be an issue. We can leave this to the user's choice.